### PR TITLE
Correctly authorize queries that use joined load and contains_eager

### PR DESCRIPTION
**Pull request by dhatch**
PR checklist:

- [x] Added changelog entry.

### CODE REVIEW

File: languages/python/sqlalchemy-oso/sqlalchemy_oso/session.py [hunk 1]
```diff
@@ -291,41 +301,32 @@ def do_orm_execute(execute_state):
...
-        for entity in entities_in_statement(execute_state.statement):
+        entities = all_entities_in_statement(execute_state.statement)
+        log(f"Authorizing entities: {entities}")
+        for entity in entities:
             # If entity is an alias, get the action for the underlying class.
             if isinstance(entity, AliasedClass):
                 action = checked_permissions.get(inspect(entity).class_)  # type: ignore
+            elif inspect(entity, False) is not None:
+                entity = inspect(entity).class_
+                action = checked_permissions.get(inspect(entity).class_)
```
[the remark this answers is not in the document]
Comment (dhatch) [reply languages/python/sqlalchemy-oso/sqlalchemy_oso/session.py]: I cleaned this up.

File: languages/python/sqlalchemy-oso/tests/test_sqlalchemy.py [hunk 2]
```diff
@@ -122,6 +122,22 @@ def test_authorize_query_multiple_types(engine, oso, fixture_data):
...
+           allow("user", "read", user: User) if user.id = 0;
+           allow("user", "read", user: User) if user.id = 1;
+           allow("all_posts", "read", _: Post);"""
+    )
+
+    session = AuthorizedSession(oso, "user", {Post: "read", User: "read"}, bind=engine)
+    authorized = session.query(User).options(joinedload(User.posts))
+    print_query(authorized)
+    print(authorized[0].posts)
+    assert len(authorized[0].posts) == 1  # Test will fail, len will be all posts
```
Comment (gw000) [inline languages/python/sqlalchemy-oso/tests/test_sqlalchemy.py:138]: It passes now, right?

Comment (dhatch) [reply languages/python/sqlalchemy-oso/tests/test_sqlalchemy.py]: Yeah, removed comment.

File: languages/python/sqlalchemy-oso/sqlalchemy_oso/sqlalchemy_utils.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+    def all_entities_in_statement(statement):
+        """
+        Get all ORM entities that will be loaded in a select statement.
+
+        The includes entities that will be loaded eagerly through relationship
+        loading: https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html#relationship-loading-with-loader-options
+        """
+        entities = get_column_entities(statement)
+        entities |= set(get_joinedload_entities(statement))
```
Comment (gw000) [inline languages/python/sqlalchemy-oso/sqlalchemy_oso/sqlalchemy_utils.py:27]: could `get_joinedload_entities` return a set too?

Comment (dhatch) [reply languages/python/sqlalchemy-oso/sqlalchemy_oso/sqlalchemy_utils.py]: Just changed.

**Review by gw000 — APPROVED** [review]
So to try and summarize what this does for posterity, it reflects on the models in the query to find which other models will be eager loaded along with them, and includes those models in the authorization for  `AuthorizedSession`. That's basically it right?

File: languages/python/sqlalchemy-oso/sqlalchemy_oso/sqlalchemy_utils.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+                        # the "path" is a tuple showing the entity/relationships
+                        # being targeted
+
+                        # TODO check for wild card.
+                        # TODO: Check whether entity is a string.
+                        yield b.path[-1].entity
+            elif hasattr(opt, "context"):
+                # these options are called Load
+                for key, loadopt in opt.context.items():
+                    if key[0] == "loader" and ("lazy", "joined") in loadopt.strategy:
```
[the remark this answers is not in the document]
Comment (dhatch) [reply languages/python/sqlalchemy-oso/sqlalchemy_oso/sqlalchemy_utils.py:146]: Yep :(

File: languages/python/sqlalchemy-oso/sqlalchemy_oso/session.py [hunk 5]
```diff
@@ -291,41 +295,25 @@ def do_orm_execute(execute_state):
...
+        logger.info(f"Authorizing entities: {entities}")
+        for entity in entities:
+            action = checked_permissions.get(entity)
 
             # If permissions map does not specify an action to authorize for entity
             # or if the specified action is `None`, deny access.
             if action is None:
+                logger.warning(f"No allowed action for entity {entity}")
                 where = with_loader_criteria(entity, expr.false(), include_aliases=True)
                 execute_state.statement = execute_state.statement.options(where)
```
[the remark this answers is not in the document]
Comment (dhatch) [reply languages/python/sqlalchemy-oso/sqlalchemy_oso/session.py:308]: Yeah, they do.

[the remark this answers is not in the document]
Comment (dhatch) [reply languages/python/sqlalchemy-oso/sqlalchemy_oso/session.py:308]: This is consistent with the current behavior, unless you're using `lazy="joined"` which would not be authorized properly (basically a bug).

File: languages/python/sqlalchemy-oso/sqlalchemy_oso/sqlalchemy_utils.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        """
+        # there are two kinds of options that both represent the same information,
+        # just in different ways.  This is largely a product of legacy options
+        # that have things like strings, i.e. joinedload("addresses").  note we
+        # aren't covering that here, which is legacy form.  you can if you want
+        # raise an exception if you detect that form here.
+
+        for opt in stmt._with_options:
+            if hasattr(opt, "_to_bind"):
+                # these options are called _UnboundLoad
```
[the remark this answers is not in the document]
Comment (dhatch) [reply languages/python/sqlalchemy-oso/sqlalchemy_oso/sqlalchemy_utils.py:134]: Yeah - should move to a public API in the future.

**Pull request merged**